index.js

```diff
@@ -196,48 +199,49 @@ AFRAME.registerComponent('teleport-controls', {
 
       // Set default status as non-hit
       this.teleportEntity.setAttribute('visible', true);
-      this.line.material.color.set(this.curveMissColor);
-      this.line.material.opacity = this.data.missOpacity;
-      this.line.material.transparent = this.data.missOpacity < 1;
       this.hitEntity.setAttribute('visible', false);
       this.hit = false;
+      this.collidedIndex = -1;
+      this.setLineMaterial(this.hit);
 
       if (this.data.type === 'parabolic') {
         v0.copy(direction).multiplyScalar(this.data.curveShootingSpeed);
-
         this.lastDrawnIndex = 0;
-        const numPoints = this.data.drawIncrementally ? this.numActivePoints : this.line.numPoints;
-        for (var i = 0; i < numPoints+1; i++) {
-          var t;
-          if (i == Math.floor(numPoints+1)){
-            t =  numPoints / (this.line.numPoints - 1);
-          }
-          else {
-            t = i / (this.line.numPoints - 1);
-          }
+        // const numPoints = this.data.drawIncrementally ? this.numActivePoints : this.line.numPoints;
+        const numPoints = this.raycastPoints.length;
+        const numSegments = numPoints - 1;
+        for (let i = 0; i < numPoints; i++) {
+          let t = i / numSegments;
           parabolicCurve(p0, v0, a, t, next);
+          this.raycastPoints[i].copy(next);
           // Update the raycaster with the length of the current segment last->next
           var dirLastNext = lastNext.copy(next).sub(last).normalize();
           this.raycaster.far = dirLastNext.length();
           this.raycaster.set(last, dirLastNext);
-
-          this.lastDrawnPoint = next;
-          this.lastDrawnIndex = i;
-          if (this.checkMeshCollisions(i, next)) { break; }
-
           last.copy(next);
-        }
-        for (var j = this.lastDrawnIndex+1; j < this.line.numPoints; j++) {
-          this.line.setPoint(j, this.lastDrawnPoint);
+
+          if (this.isMeshCollided()) {
+            this.collidedIndex = i;
+            this.hit = true;
```

Comment:
The logic in onButtonUp will have to change such that you can't teleport unless the line has reached the hitEntity (which perhaps just means that the hit is in a valid teleport location and enough time has passed since the teleport started, since we're making long-distance and short-distance teleports take the same amount of time).